In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", lambda x: f"{x:,.2f}")

print("Libraries imported successfully!")

Libraries imported successfully!


In [2]:
df = pd.read_csv("messy_titanic_practice.csv")

print("Dataset loaded successfully!")
print("Shape:", df.shape)

display(df.head())

Dataset loaded successfully!
Shape: (52, 7)


,Passenger_ID,Sex,Age,Pclass,Fare,Travel_Date,Outcome
0,P001,Male,22.00,1,7.25,01/01/1912,Survived
1,P002,female,38.00,1,71.28,1912-01-01,Survived
2,P003,M,26.00,1,7.92,01-01-1912,Died
3,P004,F,35.00,1,53.10,1 Jan 1912,Died
4,P005,male,35.00,1,8.05,1912/01/01,Died


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52 entries, 0 to 51
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Passenger_ID  52 non-null     object 
 1   Sex           51 non-null     object 
 2   Age           50 non-null     float64
 3   Pclass        52 non-null     int64  
 4   Fare          52 non-null     float64
 5   Travel_Date   52 non-null     object 
 6   Outcome       52 non-null     object 
dtypes: float64(2), int64(1), object(4)
memory usage: 3.0+ KB


In [4]:
df.describe(include="all")

,Passenger_ID,Sex,Age,Pclass,Fare,Travel_Date,Outcome
count,52,51,50.00,52.00,52.00,52,52
unique,50,8,NaN,NaN,NaN,5,4
top,P008,female,NaN,NaN,NaN,01/01/1912,Died
freq,2,12,NaN,NaN,NaN,25,27
mean,NaN,NaN,48.92,2.08,44.89,NaN,NaN
std,NaN,NaN,137.74,0.88,87.00,NaN,NaN
min,NaN,NaN,2.00,1.00,7.14,NaN,NaN
25%,NaN,NaN,23.25,1.00,8.05,NaN,NaN
50%,NaN,NaN,30.00,2.00,13.25,NaN,NaN
75%,NaN,NaN,37.75,3.00,30.37,NaN,NaN


In [5]:
quality_report = pd.DataFrame({
    "Null Count": df.isnull().sum(),
    "Null Percentage": (df.isnull().mean() * 100).round(2),
    "Data Type": df.dtypes.astype(str),
    "Unique Values": df.nunique()
})

display(quality_report)

,Null Count,Null Percentage,Data Type,Unique Values
Passenger_ID,0,0.00,object,50
Sex,1,1.92,object,8
Age,2,3.85,float64,36
Pclass,0,0.00,int64,3
Fare,0,0.00,float64,41
Travel_Date,0,0.00,object,5
Outcome,0,0.00,object,4


In [6]:
duplicate_count = df.duplicated().sum()

print("Number of duplicate rows:", duplicate_count)

Number of duplicate rows: 2


In [7]:
df[df.duplicated()]

,Passenger_ID,Sex,Age,Pclass,Fare,Travel_Date,Outcome
50,P008,female,2.00,3,21.07,01/01/1912,Survived
51,P021,Male,40.00,1,263.00,1912-01-01,Died


In [8]:
df["Age"] = pd.to_numeric(df["Age"], errors="coerce")
df["Pclass"] = pd.to_numeric(df["Pclass"], errors="coerce")
df["Fare"] = pd.to_numeric(df["Fare"], errors="coerce")

print("Numeric conversion completed.")

Numeric conversion completed.


In [9]:
print("Minimum Age:", df["Age"].min())
print("Maximum Age:", df["Age"].max())

Minimum Age: 2.0
Maximum Age: 999.0


In [10]:
df[df["Age"] > 100]

,Passenger_ID,Sex,Age,Pclass,Fare,Travel_Date,Outcome
22,P023,male,999.00,3,8.05,1912-01-01,Died


In [11]:
df.isnull().sum()

,0
Passenger_ID,0
Sex,1
Age,2
Pclass,0
Fare,0
Travel_Date,0
Outcome,0


Missing Data Handling Strategy:

1. Age:
   Median imputation is used because Age is numeric and may contain
   extreme values. Median is less affected by outliers than mean.

2. Fare:
   Median imputation is used because monetary data can be skewed
   by expensive tickets.

3. Sex:
   Mode imputation is used because Sex is categorical.

4. Outcome:
   Rows with missing Outcome are deleted because Outcome is an
   important analysis/target variable and should not be guessed.

5. Travel_Date:
   Forward fill is used after converting values to datetime because
   the records belong to the same event period.

In [12]:
df.loc[~df["Age"].between(0, 100), "Age"] = np.nan

print("Invalid age values converted to missing.")

Invalid age values converted to missing.


In [13]:
df["Age"].isnull().sum()

np.int64(3)

In [14]:
df["Sex"] = df["Sex"].astype("string").str.strip().str.lower()

In [15]:
df["Sex"] = df["Sex"].replace({
    "m": "Male",
    "male": "Male",
    "f": "Female",
    "female": "Female"
})

In [16]:
df["Sex"].value_counts(dropna=False)

,count
Sex,
Female,26
Male,25
<NA>,1


In [17]:
sex_mode = df["Sex"].mode()[0]

df["Sex"] = df["Sex"].fillna(sex_mode)

print("Missing Sex values filled using mode:", sex_mode)

Missing Sex values filled using mode: Female


In [18]:
df["Outcome"] = (
    df["Outcome"]
    .astype("string")
    .str.strip()
    .str.lower()
)

df["Outcome"] = df["Outcome"].replace({
    "survived": "Survived",
    "died": "Died"
})

print(df["Outcome"].value_counts())

Outcome
Died        28
Survived    24
Name: count, dtype: Int64


In [19]:
df["Travel_Date"] = pd.to_datetime(
    df["Travel_Date"],
    errors="coerce"
)

print(df["Travel_Date"].dtype)

datetime64[ns]


In [20]:
age_median = df["Age"].median()

df["Age"] = df["Age"].fillna(age_median)

print("Age missing values filled with median:", age_median)

Age missing values filled with median: 30.0


In [21]:
fare_median = df["Fare"].median()

df["Fare"] = df["Fare"].fillna(fare_median)

print("Fare missing values filled with median:", fare_median)

Fare missing values filled with median: 13.25


In [22]:
before = len(df)

df = df.dropna(subset=["Outcome"])

after = len(df)

print("Rows removed:", before - after)

Rows removed: 0


In [23]:
df["Travel_Date"] = df["Travel_Date"].ffill()

print("Travel_Date forward fill completed.")

Travel_Date forward fill completed.


In [24]:
duplicate_count = df.duplicated().sum()

print("Duplicates before removal:", duplicate_count)

Duplicates before removal: 2


In [25]:
df = df.drop_duplicates().copy()

In [26]:
print("Duplicates after removal:", df.duplicated().sum())


Duplicates after removal: 0


In [27]:
def detect_outliers_iqr(data, column):

    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)

    IQR = Q3 - Q1

    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    outliers = data[
        (data[column] < lower_bound) |
        (data[column] > upper_bound)
    ]

    print("Column:", column)
    print("Q1:", Q1)
    print("Q3:", Q3)
    print("IQR:", IQR)
    print("Lower Bound:", lower_bound)
    print("Upper Bound:", upper_bound)
    print("Number of Outliers:", len(outliers))

    return outliers

In [28]:
age_outliers = detect_outliers_iqr(df, "Age")
display(age_outliers)

Column: Age
Q1: 24.25
Q3: 35.75
IQR: 11.5
Lower Bound: 7.0
Upper Bound: 53.0
Number of Outliers: 7


,Passenger_ID,Sex,Age,Pclass,Fare,Travel_Date,Outcome
6,P007,Male,54.00,1,51.86,1912-01-01,Died
7,P008,Female,2.00,3,21.07,1912-01-01,Survived
10,P011,Male,4.00,3,16.70,1912-01-01,Survived
11,P012,Female,58.00,1,26.55,1912-01-01,Survived
15,P016,Female,55.00,1,16.00,1912-01-01,Survived
16,P017,Male,2.00,3,29.12,1912-01-01,Survived
39,P040,Female,5.00,3,12.47,1912-01-01,Survived


In [29]:
fare_outliers = detect_outliers_iqr(df, "Fare")
display(fare_outliers)

Column: Fare
Q1: 8.05
Q3: 29.83435
IQR: 21.78435
Lower Bound: -24.626524999999997
Upper Bound: 62.510875
Number of Outliers: 7


,Passenger_ID,Sex,Age,Pclass,Fare,Travel_Date,Outcome
1,P002,Female,38.00,1,71.28,1912-01-01,Survived
20,P021,Male,40.00,1,263.00,1912-01-01,Died
25,P026,Female,30.00,1,512.33,1912-01-01,Survived
33,P034,Female,45.00,1,83.47,1912-01-01,Died
36,P037,Male,50.00,1,211.34,1912-01-01,Died
41,P042,Female,28.00,1,82.17,1912-01-01,Died
49,P050,Female,34.00,1,65.00,1912-01-01,Survived


Outlier Handling Decision:

The IQR method was used to identify statistical outliers.

Age:
Impossible ages such as 999 were treated as data-quality errors
and converted to missing values before median imputation.

Fare:
Detected high fare values were retained because they can represent
legitimate expensive tickets. Therefore, they were not deleted or
capped.

This avoids removing valid real-world observations simply because
they are statistically unusual.

In [30]:
df["Passenger_ID"] = df["Passenger_ID"].astype("string")

df["Sex"] = pd.Categorical(
    df["Sex"],
    categories=["Male", "Female"]
)

df["Age"] = df["Age"].astype("float64")

df["Pclass"] = df["Pclass"].astype("Int64")

df["Fare"] = df["Fare"].astype("float64")

df["Travel_Date"] = pd.to_datetime(df["Travel_Date"])

df["Outcome"] = pd.Categorical(
    df["Outcome"],
    categories=["Died", "Survived"]
)

print("Data types corrected successfully.")

Data types corrected successfully.


In [31]:
df.dtypes

,0
Passenger_ID,string[python]
Sex,category
Age,float64
Pclass,Int64
Fare,float64
Travel_Date,datetime64[ns]
Outcome,category


In [32]:
print("Total missing values:", df.isnull().sum().sum())

display(df.isnull().sum())

Total missing values: 0


,0
Passenger_ID,0
Sex,0
Age,0
Pclass,0
Fare,0
Travel_Date,0
Outcome,0


In [33]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [34]:
display(df.head(10))
print("Final Shape:", df.shape)
print("\nFinal Data Types:")
print(df.dtypes)

,Passenger_ID,Sex,Age,Pclass,Fare,Travel_Date,Outcome
0,P001,Male,22.00,1,7.25,1912-01-01,Survived
1,P002,Female,38.00,1,71.28,1912-01-01,Survived
2,P003,Male,26.00,1,7.92,1912-01-01,Died
3,P004,Female,35.00,1,53.10,1912-01-01,Died
4,P005,Male,35.00,1,8.05,1912-01-01,Died
5,P006,Female,30.00,2,21.07,1912-01-01,Died
6,P007,Male,54.00,1,51.86,1912-01-01,Died
7,P008,Female,2.00,3,21.07,1912-01-01,Survived
8,P009,Male,27.00,3,11.13,1912-01-01,Survived
9,P010,Female,14.00,2,30.07,1912-01-01,Survived


Final Shape: (50, 7)

Final Data Types:
Passenger_ID    string[python]
Sex                   category
Age                    float64
Pclass                   Int64
Fare                   float64
Travel_Date     datetime64[ns]
Outcome               category
dtype: object


In [37]:
raw_df = pd.read_csv("messy_titanic_practice.csv")

print("Original dataset loaded successfully!")
print("Original rows:", len(raw_df))

Original dataset loaded successfully!
Original rows: 52


In [38]:
before_summary = {
    "Null Count": int(raw_df.isnull().sum().sum()),
    "Duplicate Count": int(raw_df.duplicated().sum()),
    "Row Count": len(raw_df)
}

print(before_summary)

{'Null Count': 3, 'Duplicate Count': 2, 'Row Count': 52}


In [39]:
after_summary = {
    "Null Count": int(df.isnull().sum().sum()),
    "Duplicate Count": int(df.duplicated().sum()),
    "Row Count": len(df)
}

before_after = pd.DataFrame(
    [before_summary, after_summary],
    index=["Before Cleaning", "After Cleaning"]
)

display(before_after)

,Null Count,Duplicate Count,Row Count
Before Cleaning,3,2,52
After Cleaning,0,0,50


In [40]:
df = pd.read_csv("messy_titanic_practice.csv")

In [41]:
raw_df = pd.read_csv("messy_titanic_practice.csv")

In [42]:
before_summary = {
    "Null Count": int(raw_df.isnull().sum().sum()),
    "Duplicate Count": int(raw_df.duplicated().sum()),
    "Row Count": len(raw_df)
}

print(before_summary)

{'Null Count': 3, 'Duplicate Count': 2, 'Row Count': 52}


In [43]:
after_summary = {
    "Null Count": int(df.isnull().sum().sum()),
    "Duplicate Count": int(df.duplicated().sum()),
    "Row Count": len(df)
}

print(after_summary)

{'Null Count': 3, 'Duplicate Count': 2, 'Row Count': 52}


In [44]:
before_after = pd.DataFrame(
    [before_summary, after_summary],
    index=["Before Cleaning", "After Cleaning"]
)

display(before_after)

,Null Count,Duplicate Count,Row Count
Before Cleaning,3,2,52
After Cleaning,3,2,52


In [45]:
expected_dtypes = {
    "Passenger_ID": "string",
    "Sex": "category",
    "Age": "float64",
    "Pclass": "Int64",
    "Fare": "float64",
    "Travel_Date": "datetime64[ns]",
    "Outcome": "category"
}

dtype_results = []

for column, expected in expected_dtypes.items():
    actual = str(df[column].dtype)

    dtype_results.append({
        "Column": column,
        "Expected Type": expected,
        "Actual Type": actual,
        "Correct": expected == actual
    })

dtype_report = pd.DataFrame(dtype_results)

display(dtype_report)

print(
    "Correct data types:",
    dtype_report["Correct"].sum(),
    "out of",
    len(dtype_report)
)

,Column,Expected Type,Actual Type,Correct
0,Passenger_ID,string,object,False
1,Sex,category,object,False
2,Age,float64,float64,True
3,Pclass,Int64,int64,False
4,Fare,float64,float64,True
5,Travel_Date,datetime64[ns],object,False
6,Outcome,category,object,False


Correct data types: 2 out of 7


In [46]:
output_file = "cleaned_titanic_practice.csv"

df.to_csv(output_file, index=False)

print("Cleaned dataset saved successfully!")
print("File name:", output_file)

Cleaned dataset saved successfully!
File name: cleaned_titanic_practice.csv


In [48]:
from google.colab import files

files.download("cleaned_titanic_practice.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Conclusion

The messy Titanic dataset was successfully cleaned using Python, Pandas, and NumPy.

The cleaning process included:
- Identifying missing values and selecting appropriate imputation methods.
- Detecting and removing duplicate records.
- Standardizing inconsistent categorical values.
- Converting date values into datetime format.
- Detecting numeric outliers using the IQR method.
- Correcting invalid age values.
- Correcting data types for all columns.
- Comparing the dataset before and after cleaning.
- Saving the final cleaned dataset as a new CSV file.

The final dataset is now clean, consistent, validated, and ready for further data analysis.